rust/agama-network/src/dbus/interfaces.rs

```diff
@@ -0,0 +1,402 @@
+use crate::{
+    action::Action,
+    error::NetworkStateError,
+    model::{Connection as NetworkConnection, Device as NetworkDevice, WirelessConnection},
+};
+use parking_lot::{MappedMutexGuard, Mutex, MutexGuard};
+use std::{
+    net::{AddrParseError, Ipv4Addr},
+    sync::{mpsc::Sender, Arc},
+};
+use uuid::Uuid;
+use zbus::{dbus_interface, zvariant::ObjectPath};
+
+/// D-Bus interface for the network devices collection
+///
+/// It offers an API to query the devices collection.
+pub struct Devices {
+    objects: Arc<Mutex<ObjectsRegistry>>,
+}
+
+impl Devices {
+    /// Creates a Devices interface object.
+    ///
+    /// * `objects`: Objects paths registry.
+    pub fn new(objects: Arc<Mutex<ObjectsRegistry>>) -> Self {
+        Self { objects }
+    }
+}
+
+#[dbus_interface(name = "org.opensuse.Agama.Network1.Devices")]
+impl Devices {
+    /// Returns the D-Bus paths of the network devices.
+    pub fn get_devices(&self) -> Vec<ObjectPath> {
+        let objects = self.objects.lock();
+        objects
+            .devices_paths()
+            .iter()
+            .filter_map(|c| ObjectPath::try_from(c.clone()).ok())
+            .collect()
+    }
+}
+
+/// D-Bus interface for a network device
+///
+/// It offers an API to query basic networking devices information (e.g., the name).
+pub struct Device {
+    device: NetworkDevice,
+}
+
+impl Device {
+    /// Creates an interface object.
+    ///
+    /// * `device`: network device.
+    pub fn new(device: NetworkDevice) -> Self {
+        Self { device }
+    }
+}
+
+#[dbus_interface(name = "org.opensuse.Agama.Network1.Device")]
+impl Device {
+    #[dbus_interface(property)]
+    pub fn name(&self) -> &str {
+        &self.device.name
+    }
+
+    #[dbus_interface(property)]
+    pub fn device_type(&self) -> u8 {
+        self.device.ty as u8
+    }
+}
+
+/// D-Bus interface for the set of connections.
+///
+/// It offers an API to query the connections collection.
+pub struct Connections {
+    actions: Arc<Mutex<Sender<Action>>>,
+    objects: Arc<Mutex<ObjectsRegistry>>,
+}
+
+impl Connections {
+    /// Creates a Connections interface object.
+    ///
+    /// * `objects`: Objects paths registry.
+    pub fn new(objects: Arc<Mutex<ObjectsRegistry>>, actions: Sender<Action>) -> Self {
+        Self {
+            objects,
+            actions: Arc::new(Mutex::new(actions)),
+        }
+    }
+}
+
+#[dbus_interface(name = "org.opensuse.Agama.Network1.Connections")]
+impl Connections {
+    /// Returns the D-Bus paths of the network connections.
+    pub fn get_connections(&self) -> Vec<ObjectPath> {
+        let objects = self.objects.lock();
+        objects
+            .connections_paths()
+            .iter()
+            .filter_map(|c| ObjectPath::try_from(c.clone()).ok())
+            .collect()
+    }
+
+    /// Adds a new network connection.
+    ///
+    /// * `name`: connection name.
+    /// * `ty`: connection type (see [crate::model::DeviceType]).
```

Comment:
to be honest I do not like much that `ty`. Why not `type` or `ctype`? so it is more readable

Reply:
`type` is a reserved word, but I am open to find a better name 😅. Why `ctype`?

Reply:
ctype as connection type acronym...btw my suggestion is then to use `type_` everywhere. It is clear that it is type and another underscore does not hurt readability..